2.0/m_rpg.cpp

```diff
@@ -0,0 +1,374 @@
+ * NARRATOR and NARRATORA were my ideas.
+ */
+
+/* $ModDesc: Provides NPC, NPCA, AMBIANCE, NARRATOR, and NARRATORA commands for use by Game Masters doing pen & paper RPGs via IRC */
+/* $ModAuthor: Naram Qashat (CyberBotX) */
+/* $ModAuthorMail: cyberbotx@cyberbotx.com */
+/* $ModDepends: core 2.0 */
+
+#include "inspircd.h"
+
+/* Removes any ! characters from a given nick */
+static std::string strip_npc_nick(const std::string &nick)
+{
+	std::string newnick;
+	unsigned len = nick.size();
+	for (unsigned x = 0; x < len; ++x)
+	{
+		char c = nick[x];
+		if (c != '!')
+			newnick += c;
+	}
+	return newnick;
+}
+
+/*
+ * NOTE: For all commands, the user in the Handle function is checked to be local or not.
+ *
+ * If they are local, then the command passed through the module's OnPreCommand and the
+ * text was set accordingly to prevent colon eating from happening. Channel is checked,
+ * and if valid, user status for being an op in the channel is checked. Assuming all that
+ * succeeds, then the command is set to the channel locally and then broadcast via ENCAP.
+ *
+ * If they are not local, then the command must've come remotely and thus is being sent
+ * directly to the handler. No channel or user checks are done, as they are assumed to
+ * have been valid on the originating server, but the text was passed via ENCAP in such
+ * a way that colon eating is not an issue and the text must be set, otherwise the
+ * previous text from a local usage will be used. Broadcasting is skipped, as it would
+ * be pretty bad to broadcast infinitely.
+ */
+
+/** Base class for /NPC and /NPCA
+ */
+class NPCx
+{
+	std::string cmdName, text;
+
+public:
+	NPCx(const std::string &cmd) : cmdName(cmd)
+	{
+	}
+
+	CmdResult Handle(const std::vector<std::string> &parameters, User *user, bool action)
+	{
+		Channel *c = ServerInstance->FindChan(parameters[0]);
+		LocalUser *localUser = IS_LOCAL(user);
+		if (localUser)
+		{
+			if (c)
+			{
+				if (!c->HasUser(user))
+				{
+					user->WriteNumeric(ERR_NOTONCHANNEL, "%s %s :You are not on that channel!", user->nick.c_str(), parameters[0].c_str());
+					return CMD_FAILURE;
+				}
+				if (c->GetPrefixValue(user) < OP_VALUE)
+				{
+					user->WriteNumeric(ERR_CHANOPRIVSNEEDED, "%s %s :You're not a channel operator", user->nick.c_str(), parameters[0].c_str());
+					return CMD_FAILURE;
+				}
+			}
+			else
+			{
+				user->WriteNumeric(ERR_NOSUCHCHANNEL, "%s %s :No such channel", user->nick.c_str(), parameters[0].c_str());
+				return CMD_FAILURE;
+			}
+		}
+		else
+			this->text = parameters[2];
+
+		/* Source is in the form of: *[nick]*!npc@[server-name] */
+		std::string npc_nick = strip_npc_nick(parameters[1]);
+		std::string npc_source = "*" + npc_nick + "*!npc@" + ServerInstance->Config->ServerName;
+
+		c->WriteChannelWithServ(npc_source, "PRIVMSG %s :%s%s%s", c->name.c_str(), action ? "\1ACTION " : "", this->text.c_str(), action ? "\1" : "");
+
+		if (localUser)
+		{
+			std::vector<std::string> params;
```

Comment:
No, because of a similar issue to why I had to implement the OnPreCommand override in the module itself. If the user types in some text and one of the words starts with a colon, it gets eaten on the broadcast.